controllers/vmi_eviction_controller.go

```diff
@@ -0,0 +1,212 @@
+	}
+
+	cluster := &clusterv1.Cluster{}
+	err := r.Get(ctx, client.ObjectKey{Namespace: clusterNS, Name: clusterName}, cluster)
+	if err != nil {
+		return nil, fmt.Errorf("can't find the cluster %s/%s; %w", clusterNS, clusterName, err)
+	}
+
+	return cluster, nil
+}
+
+// This functions drains a node from a tenant cluster.
+// The function returns 3 values:
+// * drain done - boolean
+// * retry time, or 0 if not needed
+// * error - to be returned if we want to retry
+func (r *VmiEvictionReconciler) drainNode(goctx goContext.Context, cluster *clusterv1.Cluster, nodeName string, logger logr.Logger) (bool, time.Duration, error) {
+	ctx := &context.MachineContext{Context: goctx, KubevirtCluster: &infrav1.KubevirtCluster{ObjectMeta: metav1.ObjectMeta{Namespace: cluster.Namespace, Name: cluster.Name}}}
+	kubeClient, err := r.workloadCluster.GenerateWorkloadClusterK8sClient(ctx)
+	if err != nil {
+		logger.Error(err, "Error creating a remote client while deleting Machine, won't retry")
+		return false, 0, nil
+	}
+
+	node, err := kubeClient.CoreV1().Nodes().Get(goctx, nodeName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// If an admin deletes the node directly, we'll end up here.
+			logger.Error(err, "Could not find node from noderef, it may have already been deleted")
+			return true, 0, nil
+		}
+		return false, 0, fmt.Errorf("unable to get node %q: %w", nodeName, err)
+	}
+
+	drainer := &kubedrain.Helper{
+		Client:              kubeClient,
+		Ctx:                 ctx,
+		Force:               true,
+		IgnoreAllDaemonSets: true,
+		DeleteEmptyDirData:  true,
+		GracePeriodSeconds:  -1,
+		// If a pod is not evicted in 20 seconds, retry the eviction next time the
+		// machine gets reconciled again (to allow other machines to be reconciled).
+		Timeout: 20 * time.Second,
+		OnPodDeletedOrEvicted: func(pod *corev1.Pod, usingEviction bool) {
+			verbStr := "Deleted"
+			if usingEviction {
+				verbStr = "Evicted"
+			}
+			logger.Info(fmt.Sprintf("%s pod from Node", verbStr),
+				"pod", fmt.Sprintf("%s/%s", pod.Name, pod.Namespace))
+		},
+		Out: writer{logger.Info},
+		ErrOut: writer{func(msg string, keysAndValues ...interface{}) {
+			logger.Error(nil, msg, keysAndValues...)
+		}},
+	}
+
+	if noderefutil.IsNodeUnreachable(node) {
+		// When the node is unreachable and some pods are not evicted for as long as this timeout, we ignore them.
+		drainer.SkipWaitForDeleteTimeoutSeconds = 60 * 5 // 5 minutes
+	}
+
+	if err = kubedrain.RunCordonOrUncordon(drainer, node, true); err != nil {
+		// Machine will be re-reconciled after a cordon failure.
+		logger.Error(err, "Cordon failed")
+		return false, 0, errors.Errorf("unable to cordon node %s: %v", nodeName, err)
+	}
+
+	if err = kubedrain.RunNodeDrain(drainer, node.Name); err != nil {
+		// Machine will be re-reconciled after a drain failure.
+		logger.Error(err, "Drain failed, retry in 20s", "node name", nodeName)
+		return false, 20 * time.Second, nil
```

Comment:
we need a global timeout for drain as well. Something that ensures we eventually give up attempting to drain and delete the VMI after `x` minutes. 10 minutes would likely be a pretty conservative number to use.

Reply:
an annotation on the VMI could be used to track when the drain begins. The timeout could be anchored to the start time recorded in the annotatino.